plugin-server/src/main/ingestion-queues/session-recording-v2/sessions/s3-session-batch-writer.ts

```diff
@@ -149,22 +151,101 @@ class S3SessionBatchFileWriter implements SessionBatchFileWriter {
     private generateKey(): string {
         const timestamp = Date.now()
         const suffix = randomBytes(8).toString('hex')
-        return `${this.prefix}/${timestamp}-${suffix}`
+
+        // TODO: Remove this code path when all teams have adopted the new retention mechanism
+        if (this.retentionPeriod === 'legacy') {
+            return `session_recording_batches/${timestamp}-${suffix}`
+        }
+
+        return `${this.prefix}/${this.retentionPeriod}/${timestamp}-${suffix}`
     }
 }
 
 export class S3SessionBatchFileStorage implements SessionBatchFileStorage {
+    private batchWriters: { [key in RetentionPeriod]: SessionBatchFileWriter | null }
+    private activeBatch: boolean
+
     constructor(
         private readonly s3: S3Client,
         private readonly bucket: string,
         private readonly prefix: string,
         private readonly timeout: number = 5000
     ) {
+        this.batchWriters = ValidRetentionPeriods.reduce(
+            (writers, retentionPeriod) => {
+                writers[retentionPeriod] = null
+                return writers
+            },
+            {} as { [key in RetentionPeriod]: SessionBatchFileWriter | null }
+        )
```

Comment:
is this equivalent?
```suggestion
	this.resetWriters()
```

(although maybe TS will complain batchWriters isn't init'd)